community/bolt/src/main/java/org/neo4j/bolt/v1/runtime/BoltChannelAutoReadLimiter.java

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright (c) 2002-2018 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.bolt.v1.runtime;
+
+import io.netty.channel.Channel;
+
+import java.util.Collection;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.neo4j.logging.Log;
+import org.neo4j.unsafe.impl.internal.dragons.FeatureToggles;
+
+import static java.util.Objects.requireNonNull;
+
+public class BoltChannelAutoReadLimiter implements BoltWorkerQueueMonitor
+{
+    protected static final String LOW_WATERMARK_NAME = "low_watermark";
+    protected static final String HIGH_WATERMARK_NAME = "high_watermark";
+
+    private static final int defaultLowWatermark = FeatureToggles.getInteger( BoltChannelAutoReadLimiter.class, LOW_WATERMARK_NAME, 100 );
+    private static final int defaultHighWatermark = FeatureToggles.getInteger( BoltChannelAutoReadLimiter.class, HIGH_WATERMARK_NAME, 300 );
+
+    private final AtomicInteger queueSize = new AtomicInteger( 0 );
```

Comment:
Can this field have type `int`? I think it is only accessed from a `RunnableBoltWorker` thread

Reply:
`RunnableBoltWorker#queue` is invoked by netty IO threads, so I think it would be better to keep it as an `AtomicInteger`. Am I missing some point?

Reply:
You are right, never mind me :)